main/test/ca/uwaterloo/flix/library/TestDelayMap.flix

```diff
@@ -112,4 +112,471 @@ namespace TestDelayMap {
+
+    @test
+    def mapImpure04(): Bool & Impure =
+        Map#{2 -> -1, 5 -> 15, 11 -> -9} |> Map.toDelayMap |> DelayMap.map(v -> 3 * v as & Impure) |> DelayMap.toMap == Map#{2 -> -3, 5 -> 45, 11 -> -27}
+
+    @test
+    def mapImpure05(): Bool & Impure =
+        Map#{2 -> -1, 5 -> 15, 11 -> -9, 8 -> 8} |> Map.toDelayMap |> DelayMap.map(v -> 3 * v as & Impure) |> DelayMap.toMap == Map#{2 -> -3, 5 -> 45, 11 -> -27, 8 -> 24}
+
+
+    /////////////////////////////////////////////////////////////////////////////
+    // map map                                                                 //
+    /////////////////////////////////////////////////////////////////////////////
+
+    @test
+    def mapMap01(): Bool =
+        Map#{2 -> -1, 5 -> 15, 11 -> -9} |>
+            Map.toDelayMap |>
+            DelayMap.map(v -> 3 * v) |>
+            DelayMap.map(v -> 3 * v) |>
+            DelayMap.toMap == Map#{2 -> -9, 5 -> 135, 11 -> -81}
+
+    @test
+    def mapMap02(): Bool & Impure =
+        Map#{2 -> -1, 5 -> 15, 11 -> -9} |>
+            Map.toDelayMap |>
+            DelayMap.map(v -> 3 * v as & Impure) |>
+            DelayMap.map(v -> 3 * v) |>
+            DelayMap.toMap == Map#{2 -> -9, 5 -> 135, 11 -> -81}
+
+    @test
+    def mapMap03(): Bool & Impure =
+        Map#{2 -> -1, 5 -> 15, 11 -> -9} |>
+            Map.toDelayMap |>
+            DelayMap.map(v -> 3 * v) |>
+            DelayMap.map(v -> 3 * v as & Impure) |>
+            DelayMap.toMap == Map#{2 -> -9, 5 -> 135, 11 -> -81}
+
+    @test
+    def mapMap04(): Bool & Impure =
+        Map#{2 -> -1, 5 -> 15, 11 -> -9} |>
+            Map.toDelayMap |>
+            DelayMap.map(v -> 3 * v as & Impure) |>
+            DelayMap.map(v -> 3 * v as & Impure) |>
+            DelayMap.toMap == Map#{2 -> -9, 5 -> 135, 11 -> -81}
+
+
+    /////////////////////////////////////////////////////////////////////////////
+    // map (fusion)                                                            //
+    /////////////////////////////////////////////////////////////////////////////
+
+    @test
+    def mapFusion01(): Bool & Impure =
+        let l = ref Nil;
+        ((1, 1) :: (2, 2) :: (3, 3) :: Nil) |> List.toDelayMap |>
+        DelayMap.map(v -> { l := "a" :: deref l; v }) |>
+        DelayMap.map(v -> { l := "b" :: deref l; v });
+        List.reverse(deref l) == ("a" :: "a" :: "a" :: "b" :: "b" :: "b" :: Nil)
+
+    @test
+    def mapFusion02(): Bool & Impure =
+        let l = ref Nil;
+        let m = ((1, 1) :: (2, 2) :: (3, 3) :: Nil) |> List.toDelayMap |>
+        DelayMap.map(v -> { l := "a" :: deref l; v } as & Pure) |>
+        DelayMap.map(v -> { l := "b" :: deref l; v } as & Pure);
+        let listBeforeEvaluation = deref l;
+        let _ = DelayMap.toList(m);
+        listBeforeEvaluation == Nil and List.reverse(deref l) == "a" :: "b" :: "a" :: "b" :: "a" :: "b" :: Nil
```

Comment:
I think this check is very wierd. I would just focus on the second part and forget about the listBefore whatever.

If you really want that, I would have a separate test with a ref Boolean.